src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                    bool asnReadSuccessful;
 
-                    int size = (int)result[0];
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
-
-                    PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = pageControl;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.1504")
-                {
-                    // The control is an AsqControl.
-                    object[] o = BerConverter.Decode("{e}", value);
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    if (controls[i].Type == "1.2.840.113556.1.4.319")
+                    {
+                        // The control is a PageControl, as described in RFC 2696.
+                        byte[] cookie = Array.Empty<byte>();
 
-                    int result = (int)o[0];
-                    AsqResponseControl asq = new AsqResponseControl(result, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = asq;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.841")
-                {
-                    // The control is a DirSyncControl.
-                    object[] o = BerConverter.Decode("{iiO}", value);
-                    Debug.Assert(o != null && o.Length == 3);
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
 
-                    int moreData = (int)o[0];
-                    int count = (int)o[1];
-                    byte[] dirsyncCookie = (byte[])o[2];
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int size, out int bytesConsumed);
+                        ThrowUnless(asnReadSuccessful);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
 
-                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, (moreData == 0 ? false : true), count, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = dirsync;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.474")
-                {
-                    // The control is a SortControl.
-                    int result;
-                    string attribute = null;
-                    object[] o = BerConverter.TryDecode("{ea}", value, out bool decodeSucceeded);
+                        // If present, the remaining bytes in the control are expected to be an octet string.
+                        if (!asnSpan.IsEmpty)
```

Comment:
This looks like it's declaring the `cookie` value to be OPTIONAL; but RFC 2696 does not describe it as being OPTIONAL.

The existing BerConverter.Decode also doesn't seem to support it being optional:

```C#
BerConverter.Decode("{iO}", new byte[] { 0x30, 0x05, 0x02, 0x01, 0x03, 0x04, 0x00 });
```

succeeds, but

```C#
BerConverter.Decode("{iO}", new byte[] { 0x30, 0x03, 0x02, 0x01, 0x03 });
```

(omitting the empty OCTET STRING value) throws a `BerConversionException`.

Since neither the existing code nor the referenced RFC mention this value as being optional, it should not be treated as such here.